### ISSUE TITLE
on_failure_callback does not seem to fire on pod deletion/eviction

### ISSUE THREAD

**Issue by george-zubrienko** [apache/airflow#14422]
**Apache Airflow version**: 2.0.1

**Kubernetes version (if you are using kubernetes)** (use `kubectl version`): 1.16.x

**Environment**: KubernetesExecutor with single scheduler pod

**What happened**: On all previous versions we used (from 1.10.x to 2.0.0), evicting or deleting a running task pod triggered the `on_failure_callback` from `BaseOperator`. We use this functionality quite a lot to detect eviction and provide work carry-over and automatic task clear.

We've recently updated our dev environment to 2.0.1 and it seems that now `on_failure_callback` is only fired when pod completes naturally, i.e. not evicted / deleted with kubectl

Everything looks the same on task log level when pod is removed with `kubectl delete pod...`:

```
Received SIGTERM. Terminating subprocesses
Sending Signals.SIGTERM to GPID 16
Received SIGTERM. Terminating subprocesses.
Task received SIGTERM signal
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/airflow/models/taskinstance.py", line 1112, in _run_raw_task
    self._prepare_and_execute_task_with_callbacks(context, task)
  File "/usr/local/lib/python3.7/site-packages/airflow/models/taskinstance.py", line 1285, in _prepare_and_execute_task_with_callbacks
    result = self._execute_task(context, task_copy)
  File "/usr/local/lib/python3.7/site-packages/airflow/models/taskinstance.py", line 1315, in _execute_task
    result = task_copy.execute(context=context)
  File "/usr/local/lib/python3.7/site-packages/li_airflow_common/custom_operators/li_operator.py", line 357, in execute
    self.operator_task_code(context)
  File "/usr/local/lib/python3.7/site-packages/li_airflow_common/custom_operators/mapreduce/yarn_jar_operator.py", line 62, in operator_task_code
    ssh_connection=_ssh_con
  File "/usr/local/lib/python3.7/site-packages/li_airflow_common/custom_operators/mapreduce/li_mapreduce_cluster_operator.py", line 469, in watch_application
    existing_apps=_associated_applications.keys()
  File "/usr/local/lib/python3.7/site-packages/li_airflow_common/custom_operators/mapreduce/li_mapreduce_cluster_operator.py", line 376, in get_associated_application_info
    logger=self.log
  File "/usr/local/lib/python3.7/site-packages/li_airflow_common/custom_operators/mapreduce/yarn_api/yarn_api_ssh_client.py", line 26, in send_request
    _response = requests.get(request)
  File "/usr/local/lib/python3.7/site-packages/requests/api.py", line 76, in get
    return request('get', url, params=params, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/requests/api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.7/site-packages/requests/sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/usr/local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 677, in urlopen
    chunked=chunked,
  File "/usr/local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 392, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/local/lib/python3.7/http/client.py", line 1277, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/local/lib/python3.7/http/client.py", line 1323, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.7/http/client.py", line 1272, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.7/http/client.py", line 1032, in _send_output
    self.send(msg)
  File "/usr/local/lib/python3.7/http/client.py", line 972, in send
    self.connect()
  File "/usr/local/lib/python3.7/site-packages/urllib3/connection.py", line 187, in connect
    conn = self._new_conn()
  File "/usr/local/lib/python3.7/site-packages/urllib3/connection.py", line 160, in _new_conn
    (self._dns_host, self.port), self.timeout, **extra_kw
  File "/usr/local/lib/python3.7/site-packages/urllib3/util/connection.py", line 74, in create_connection
    sock.connect(sa)
  File "/usr/local/lib/python3.7/site-packages/airflow/models/taskinstance.py", line 1241, in signal_handler
    raise AirflowException("Task received SIGTERM signal")
airflow.exceptions.AirflowException: Task received SIGTERM signal
Marking task as FAILED. dag_id=mock_dag_limr, task_id=SetupMockScaldingDWHJob, execution_date=20190910T000000, start_date=20210224T162811, end_date=20210224T163044
Process psutil.Process(pid=16, status='terminated', exitcode=1, started='16:28:10') (16) terminated with exit code 1
```

But `on_failure_callback` is not triggered. For simplicity, let's assume the callback does this:

```
def act_on_failure(context):
    send_slack_message(
        message=f"{context['task_instance_key_str']} fired failure callback",
        channel=get_stored_variable('slack_log_channel')
    )

def get_stored_variable(variable_name, deserialize=False):
    try:
        return Variable.get(variable_name, deserialize_json=deserialize)
    except KeyError:
        if os.getenv('PYTEST_CURRENT_TEST'):
            _root_dir = str(Path(__file__).parent)
            _vars_path = os.path.join(_root_dir, "vars.json")
            _vars_json = json.loads(open(_vars_path, 'r').read())
            if deserialize:
                return _vars_json.get(variable_name, {})
            else:
                return _vars_json.get(variable_name, "")
        else:
            raise

def send_slack_message(message, channel):
    _web_hook_url = get_stored_variable('slack_web_hook')
    post = {
        "text": message,
        "channel": channel
    }

    try:
        json_data = json.dumps(post)
        req = request.Request(
            _web_hook_url,
            data=json_data.encode('ascii'),
            headers={'Content-Type': 'application/json'}
        )

        request.urlopen(req)
    except request.HTTPError as em:
        print('Failed to send slack messsage to the hook {hook}: {msg}, request: {req}'.format(
            hook=_web_hook_url,
            msg=str(em),
            req=str(post)
        ))
```

Scheduler logs related to this event:

```
21-02-24 16:33:04,968] {kubernetes_executor.py:147} INFO - Event: mockdaglimrsetupmocksparkdwhjob.791032759a764d8bae66fc7bd7ab2db3 had an event of type MODIFIED
[2021-02-24 16:33:04,968] {kubernetes_executor.py:202} INFO - Event: mockdaglimrsetupmocksparkdwhjob.791032759a764d8bae66fc7bd7ab2db3 Pending
[2021-02-24 16:33:04,979] {kubernetes_executor.py:147} INFO - Event: mockdaglimrsetupmocksparkdwhjob.791032759a764d8bae66fc7bd7ab2db3 had an event of type DELETED
[2021-02-24 16:33:04,979] {kubernetes_executor.py:197} INFO - Event: Failed to start pod mockdaglimrsetupmocksparkdwhjob.791032759a764d8bae66fc7bd7ab2db3, will reschedule
[2021-02-24 16:33:05,406] {kubernetes_executor.py:354} INFO - Attempting to finish pod; pod_id: mockdaglimrsetupmocksparkdwhjob.791032759a764d8bae66fc7bd7ab2db3; state: up_for_reschedule; annotations: {'dag_id': 'mock_dag_limr', 'task_id': 'SetupMockSparkDwhJob', 'execution_date': '2019-09-10T00:00:00+00:00', 'try_number': '9'}
[2021-02-24 16:33:05,419] {kubernetes_executor.py:528} INFO - Changing state of (TaskInstanceKey(dag_id='mock_dag_limr', task_id='SetupMockSparkDwhJob', execution_date=datetime.datetime(2019, 9, 10, 0, 0, tzinfo=tzlocal()), try_number=9), 'up_for_reschedule', 'mockdaglimrsetupmocksparkdwhjob.791032759a764d8bae66fc7bd7ab2db3', 'airflow', '173647183') to up_for_reschedule
[2021-02-24 16:33:05,422] {scheduler_job.py:1206} INFO - Executor reports execution of mock_dag_limr.SetupMockSparkDwhJob execution_date=2019-09-10 00:00:00+00:00 exited with status up_for_reschedule for try_number 9
```

However task stays in failed state (not what scheduler says)

When pod completes on its own (fails, exits with 0), callbacks are triggered correctly

**What you expected to happen**: `on_failure_callback` is called regardless of how pod exists, including SIGTERM-based interruptions: pod eviction, pod deletion

 Not sure really. We believe this code is executed since we get full stack trace
https://github.com/apache/airflow/blob/2.0.1/airflow/models/taskinstance.py#L1149

But then it is unclear why `finally` clause here does not run:
https://github.com/apache/airflow/blob/master/airflow/models/taskinstance.py#L1422

**How to reproduce it**:

With Airflow 2.0.1 running KubernetesExecutor, execute `kubectl delete ...` on any running task pod. Task operator should define `on_failure_callback`. In order to check that it is/not called, send data from it to any external logging system

**Anything else we need to know**:
Problem is persistent and only exists in 2.0.1 version

**Comment by ephraimbuddy** [comment]
I have investigated this and fixed a solution but I'm not sure if we want it.

Basically when a pod is deleted, it sends a SIGTERM signal to the task which gets killed while being prepared for execution. 

Now, on_failure_callback is only called when tasks finishes executing. 
My solution was to check if task has on_failure_callback when a SIGTERM is received and then execute the on_failure_callback.

I'm skeptical about adding it because task was killed while being prepared for execution. 
Does it make sense to add this? @kaxil @ashb

**Comment by george-zubrienko** [comment]
Would it make sense to add `on_kill_callback` instead of touching current logic for `on_failure_callback`? It feels a bit off that we treat "kill" as failure in terms of callbacks. Mainly due to the fact that task kill usually has different post-processing applied to it. There is a `on_kill` on `BaseOperator`, but that one is not a callback function, but rather a pre-task-kill call.

**Comment by kaxil** [comment]
I think this happened in 2.0.1 mainly because of the following traceL

When you delete the POD, the KubernetesExecutor executes the following:

1)
https://github.com/apache/airflow/blob/beb8af5ac6c438c29e2c186145115fb1334a3735/airflow/executors/kubernetes_executor.py#L195-L200

i.e. it tried to reschedule your POD as evident by the logs in the Issue description too.

2) 

Which then executes the following and puts the TaskInstance Key to `result_queue`:

https://github.com/apache/airflow/blob/beb8af5ac6c438c29e2c186145115fb1334a3735/airflow/executors/kubernetes_executor.py#L350-L359

3)

The TI is then marked with state `RESCHEDULE` (atleast according to executor events) in:

https://github.com/apache/airflow/blob/beb8af5ac6c438c29e2c186145115fb1334a3735/airflow/executors/kubernetes_executor.py#L522-L542

(4)
Now all the above 3 events were happening from `KubernetesExecutor` point of view.

At the same time when the POD was killed (sent SIGTERM), the Task Pod receives the SIGTERM and executes the following call since we override SIGTERM call and raises `AirflowException` (which matches your logs and stacktrace):

https://github.com/apache/airflow/blob/beb8af5ac6c438c29e2c186145115fb1334a3735/airflow/models/taskinstance.py#L1238-L1241

This `AirflowException` is handled here:

https://github.com/apache/airflow/blob/beb8af5ac6c438c29e2c186145115fb1334a3735/airflow/models/taskinstance.py#L1142-L1150

which then calls inside `handle_failure`

https://github.com/apache/airflow/blob/beb8af5ac6c438c29e2c186145115fb1334a3735/airflow/models/taskinstance.py#L1484-L1490

which does not run a failure_callback. This bug might have been introduced in https://github.com/apache/airflow/commit/efe163a1fddfd66fa402231906e96733efddf8af where we moved running callbacks in `LocalTaskJob`:

https://github.com/apache/airflow/blob/beb8af5ac6c438c29e2c186145115fb1334a3735/airflow/jobs/local_task_job.py#L123-L126

https://github.com/apache/airflow/blob/beb8af5ac6c438c29e2c186145115fb1334a3735/airflow/jobs/local_task_job.py#L144-L153

I don't see `LocalTaskJob` exit logs in your trace so I am not sure why that happened.

Secondly, recently @jedcunningham changed (3) where we mark task as RESCHEDULED to FAILED in https://github.com/apache/airflow/pull/14810 -- which means atleast the logging around that will be taken care off, we still need to investigate on why LocalTaskJob was not executed.

**Comment by ephraimbuddy** [comment]
Thanks @kaxil, I'll update my PR

**Comment by jedcunningham** [comment]
(1-3) only comes into play before the container actually starts so there is nothing to signal (and subsequently no way for the worker to run `on_failure_callback`), otherwise the I think the analysis of (4) is on the right track.

**Comment by kaxil** [comment]
> (1-3) only comes into play before the container actually starts so there is nothing to signal (and subsequently no way for the worker to run `on_failure_callback`), otherwise the I think the analysis of (4) is on the right track.

(1) will get executed when a task is deleted no? That matches the logs from OP that says the Task is marked as Reschedule from Kubernetes Executor evens point of view. Or are you just talking about the `on_failure_callback` ?

**Comment by jedcunningham** [comment]
> (1) will get executed when a task is deleted no?

Only when a _pending_ pod is deleted. Once it is running it won't fall into that reschedule logic. I don't quite know how the OP was able to get both the traceback from the worker and the reschedule logic from the same run. I think the scheduler log is the piece that doesn't fit and must have been grabbed from deleting a pending pod, not a running one.

**Comment by ephraimbuddy** [comment]
The fix I provided resolved it but I'm close to getting another one out.

When the pod is killed, the task instance state is 'running' and AirflowException is raised too, what I'm doing now is to set this to failed and call the `handle_failure_with_callback` method.

**Comment by houqp** [comment]
Here is what I think the trace is. Every pod starts with the local task job (monitoring/manager task), which runs the actual task code (`_run_raw_task`) in a subprocess. When the pod is force killed, the local task job will receive the SIGTERM first:

https://github.com/apache/airflow/blob/d7bc217957b65471ca5f2e259bba15c71a2b0c41/airflow/jobs/local_task_job.py#L75-L79

This line prints the first `Received SIGTERM. Terminating subprocesses` log output. Then in `self.on_kill`, it sends another SIGTERM to the subprocess that runs that actual task, and which hits the following code:

This line handler prints the second `Received SIGTERM. Terminating subprocesses` log output.

https://github.com/apache/airflow/blob/d7bc217957b65471ca5f2e259bba15c71a2b0c41/airflow/models/taskinstance.py#L1263-L1266

The design in https://github.com/apache/airflow/pull/10917 is to move all success/failure callback invocations into local task job so we can handle https://github.com/apache/airflow/issues/11086. We can't rely on the actual task process (`_run_raw_task`) to execute the callback because that process itself runs arbitrary user code and can be force killed by the kernel due to OOM or it could crash by itself due to other bugs. The only safe way to make sure we execute callbacks no matter how the task subprocess exits is to execute them from the monitoring/manager task (local task job), which doesn't run user code so we can tightly control its behavior and memory foot print.

I think the source of this bug is because local job task only calls `self.on_kill` in the SIGTERM handler, which is only responsible for killing the task subprocess using SIGTERM:

https://github.com/apache/airflow/blob/d7bc217957b65471ca5f2e259bba15c71a2b0c41/airflow/jobs/local_task_job.py#L155-L157

Here is a potential fix. In local task job's `signal_handler` after we called `self.on_kill`, we can explicitly call `self.handle_task_exit`, which will check the task state and execute callbacks accordingly.

**Comment by ephraimbuddy** [comment]
Thanks @houqp, I will give it a try

**Comment by kaxil** [comment]
@houqp This is what I had suggested on call with @ephraimbuddy few days back to add `self.handle_task_exit(1)`

```diff
  def signal_handler(signum, frame): 
      """Setting kill signal handler""" 
      self.log.error("Received SIGTERM. Terminating subprocesses") 
      self.on_kill() 
+    self.handle_task_exit(1)
      raise AirflowException("LocalTaskJob received SIGTERM signal") 
```

However, that didn't work coz `_run_finished_callback`` didn't run anything since the Task state was still RUNNING. We should probably add `self.task_instance.state=State.FAILED` in `handle_task_exit` if `exit_code != 1`. WDYT  @houqp  @ephraimbuddy  ?

https://github.com/apache/airflow/blob/e7c642ba2a79ea13d6ef84b78242f6c313cd3457/airflow/jobs/local_task_job.py#L144-L153

https://github.com/apache/airflow/blob/e7c642ba2a79ea13d6ef84b78242f6c313cd3457/airflow/models/taskinstance.py#L1355-L1378

**Comment by ephraimbuddy** [comment]
@kaxil , Yes. We both tried it and it didn't work because the task was still in the RUNNING state. I have tried marking the task as Failed when AirflowException is raised and also tried a bunch of other things. I'm still on it but I think handling this the way we handle email_on_failure should fix it which is what my PR did. 

I'm yet to test sending sigkill but will test it soon on my PR and I will also try this your suggestion @kaxil

**Comment by houqp** [comment]
Interesting, I was expecting the second SIGTERM would have resulted in the task subprocess to set its own state in through `handle_failure` because `self.on_kill` calls `self.task_runner.terminate()`, which is supposed to wait for the subprocess to exit:

https://github.com/apache/airflow/blob/e7c642ba2a79ea13d6ef84b78242f6c313cd3457/airflow/task/task_runner/standard_task_runner.py#L108-L117

>  We should probably add self.task_instance.state=State.FAILED in handle_task_exit if exit_code != 1. WDYT @houqp @ephraimbuddy ?

I think we should do this as an extra safe guard because in rare cases, the task sub process could crash any time before it updates its own task state. However, I think the state update logic should be set with an extra condition:

```python
if self.task_instance.state not in State.finished():
    self.task_instance.state = State.FAILED
```

Because the task could have exited successfully right after the pod is getting killed but before local task job executes `handle_task_exit`.

**Comment by ephraimbuddy** [comment]
Hi @houqp , I tried your suggestion with that of @kaxil but they didn't work.

This is another solution that worked:
<details><summary>Changes diff</summary>
<p>

```diff
diff --git a/airflow/jobs/local_task_job.py b/airflow/jobs/local_task_job.py
index d68bfc722..6d3feb72a 100644
--- a/airflow/jobs/local_task_job.py
+++ b/airflow/jobs/local_task_job.py
@@ -76,7 +76,6 @@ class LocalTaskJob(BaseJob):
             """Setting kill signal handler"""
             self.log.error("Received SIGTERM. Terminating subprocesses")
             self.on_kill()
-            raise AirflowException("LocalTaskJob received SIGTERM signal")
 
         # pylint: enable=unused-argument
         signal.signal(signal.SIGTERM, signal_handler)
@@ -191,9 +190,16 @@ class LocalTaskJob(BaseJob):
             if ti.state == State.SUCCESS:
                 error = None
             else:
+                if ti.state == State.RUNNING:
+                    ti.state = State.FAILED
+                    session.merge(ti)
+                    session.commit()
                 # if ti.state is not set by taskinstance.handle_failure, then
                 # error file will not be populated and it must be updated by
                 # external source suck as web UI
-                error = self.task_runner.deserialize_run_error() or "task marked as failed externally"
+                try:
+                    error = self.task_runner.deserialize_run_error() or "task marked as failed externally"
+                except ValueError:
+                    error = 'task killed by signal'
             ti._run_finished_callback(error=error)  # pylint: disable=protected-access
             self.terminating = True
diff --git a/tests/jobs/test_local_task_job.py b/tests/jobs/test_local_task_job.py
index 7f9445c62..56c5d6b8b 100644
--- a/tests/jobs/test_local_task_job.py
+++ b/tests/jobs/test_local_task_job.py
@@ -29,6 +29,7 @@ import pytest
 
 from airflow import settings
 from airflow.exceptions import AirflowException, AirflowFailException
+from airflow.executors.kubernetes_executor import KubernetesExecutor
 from airflow.executors.sequential_executor import SequentialExecutor
 from airflow.jobs.local_task_job import LocalTaskJob
 from airflow.models.dag import DAG
@@ -495,6 +496,70 @@ class TestLocalTaskJob(unittest.TestCase):
         assert task_terminated_externally.value == 1
         assert not process.is_alive()
 
+    def test_process_kill_call_on_failure_callback(self):
+        """
+        Test that ensures that where a task is killed with sigterm
+        on_failure_callback gets executed
+        """
+        # use shared memory value so we can properly track value change even if
+        # it's been updated across processes.
+        failure_callback_called = Value('i', 0)
+        task_terminated_externally = Value('i', 1)
+        shared_mem_lock = Lock()
+
+        def failure_callback(context):
+            with shared_mem_lock:
+                failure_callback_called.value += 1
+            assert context['dag_run'].dag_id == 'test_mark_failure'
+
+        dag = DAG(dag_id='test_mark_failure', start_date=DEFAULT_DATE, default_args={'owner': 'owner1'})
+
+        def task_function(ti):
+            # pylint: disable=unused-argument
+            time.sleep(60)
+            # This should not happen -- the state change should be noticed and the task should get killed
+            with shared_mem_lock:
+                task_terminated_externally.value = 0
+            raise Exception
+
+        task = PythonOperator(
+            task_id='test_on_failure',
+            python_callable=task_function,
+            on_failure_callback=failure_callback,
+            dag=dag,
+        )
+
+        session = settings.Session()
+
+        dag.clear()
+        dag.create_dagrun(
+            run_id="test",
+            state=State.RUNNING,
+            execution_date=DEFAULT_DATE,
+            start_date=DEFAULT_DATE,
+            session=session,
+        )
+        ti = TaskInstance(task=task, execution_date=DEFAULT_DATE)
+        ti.refresh_from_db()
+        job1 = LocalTaskJob(task_instance=ti, ignore_ti_state=True, executor=KubernetesExecutor())
+        job1.task_runner = StandardTaskRunner(job1)
+
+        settings.engine.dispose()
+        process = multiprocessing.Process(target=job1.run)
+        process.start()
+
+        for _ in range(0, 10):
+            ti.refresh_from_db()
+            if ti.state == State.RUNNING:
+                break
+            time.sleep(0.2)
+        assert ti.state == State.RUNNING
+        os.kill(ti.pid, 15)
+        process.join(timeout=10)
+        assert failure_callback_called.value == 1
+        assert task_terminated_externally.value == 1
+        assert not process.is_alive()
+
 
 @pytest.fixture()
 def clean_db_helper():

```
</p>
</details>
Also, I tested the above solution as well as my PR(#15172) on #11086 and both worked fine.

Here's #11086 modified dag to work on 2.0+:

<details><summary>DAG CODE</summary>
<p>

```python
from datetime import datetime

import numpy as np

from airflow import DAG
from airflow.api.common.experimental.trigger_dag import trigger_dag
from airflow.operators.python import PythonOperator


def on_failure_callback(context):
    print("===IN ON FAILURE CALLBACK===")
    print("Triggering another run of the task")
    trigger_dag("OOM_test_follower")


def high_memory_task():
    l = []
    iteration = 0
    while True:
        print(f"Iteration: {iteration}")
        l.append(np.random.randint(1_000_000, size=(1000, 1000, 100)))
        iteration += 1


def failure_task():
    raise ValueError("whoops")


def print_context(ds, **kwargs):
    print("This DAG was launched by the failure callback")
    print(ds)
    print(kwargs)


dag = DAG(
    dag_id="OOM_test",
    schedule_interval=None,
    catchup=False,
    default_args={
        "owner": "madison.bowden",
        "start_date": datetime(year=2019, month=7, day=1),
        "email": "your-email",
    },
)

with dag:

    PythonOperator(
        task_id="oom_task",
        python_callable=high_memory_task,
        on_failure_callback=on_failure_callback,
    )

failure_dag = DAG(
    dag_id="Failure_test",
    schedule_interval=None,
    catchup=False,
    default_args={
        "owner": "madison.bowden",
        "start_date": datetime(year=2019, month=7, day=1),
        "email": "your-email",
    },
)

with failure_dag:

    PythonOperator(
        task_id="failure_task",
        python_callable=failure_task,
        on_failure_callback=on_failure_callback,
    )

dag_follower = DAG(
    dag_id="OOM_test_follower",
    schedule_interval=None,
    catchup=False,
    default_args={
        "owner": "madison.bowden",
        "start_date": datetime(year=2019, month=7, day=1),
        "email": "your-email",
    },
)

with dag_follower:

    PythonOperator(
        task_id="oom_task_failure", python_callable=print_context
    )

```
</p>
</details>

On testing, it looks like we can't capture SIGKILL : https://docs.python.org/3/library/signal.html#signal.SIGKILL. I will appreciate any guidance for this.

Also, I'm preferring the solution on the PR(#15172) because it works the way `email_on_failure` works. 
While this later solution works, I feel it's not appropriate.

Let me know what you think. Thanks

**Comment by houqp** [comment]
@ephraimbuddy after looking more into the code, the initial change I proposed indeed will not work due to wrong assumptions. Here are couple things you need to do to expand it into a working implementation:

* `self.task_instance.state = State.FAILED` only sets the state in memory, so you need to commit the change into database as well using `self.task_instance.set_state(State.FAILED)`
* before we do the `self.task_instance.state not in State.finished` check, we need to call `self.task_instance.refresh_from_db()` to refresh the state from DB
* `self.on_kill` calls `self.task_runner.on_finish()`, which closes `self.task_runner._error_file`. However, in `self.handle_task_exit`, we call `self.task_runner.deserialize_run_error()`, which tries to read `self.task_runner._error_file` again, this will cause `ValueError: seek of closed file` error. We need to skip the `deserialize_run_error()` call in case of sigterm since the sub task process won't have the chance to write to error file anyway. In fact, we can skip `handle_task_exit` entirely and just call `self.task_instance._run_finished_callback` directly within the signal handler.

I pushed a simple implementation at https://github.com/apache/airflow/commit/e508300ef7f50434553d9b61e872316649c93df4 for reference.

SIGKILL is not catchable by design just like OOM kills sent by the kernel, that's why I recommended using sigkill in tests to simulate failure cases caused by uncatchable errors like OOM. But the local task job should be able to handle it because it will check the process return code and set the task state accordingly (currently not implemented). If we execute callbacks from within taskinstance process, then those callbacks will never be executed on uncatchable errors like OOM.

**Comment by ephraimbuddy** [comment]
Thanks @houqp for the insights! Really appreciate. Now I'm working on the SIGKILL and will keep you posted

**Comment by houqp** [comment]
awesome, btw, I think the SIGKILL/OOM issue is something different from the bug that you fixed in #15172 , so we can work on that as a separate task/PR as well to get #15172 merged sooner.

**Comment by ephraimbuddy** [comment]
Cool, I agree with you. I debugged with @ashb on the SIGKILL and found interesting things. It should be a separate PR indeed
